crates/sui-source-validation-service/src/lib.rs

```diff
@@ -388,9 +395,32 @@ pub async fn watch_for_upgrades(
         .await?;
 
     info!("Listening for upgrades...");
-    let result = subscription.next().await;
-    info!("Saw upgrade txn: {:?}", result);
-    Ok(())
+    loop {
+        let result: Option<Result<SuiTransactionBlockEffects, _>> = subscription.next().await;
+        match result {
+            Some(Ok(result)) => {
+                // We see an upgrade transaction. Clear all sources since all of part of these may now be invalid.
+                // Currently we need to restart the server within some time delta of this observation to resume
+                // returning source. Restarting revalidates the latest release sources per repositories in the config file.
+                // Restarting is a manual side-effect outside of this server because we need to ensure that sources in the
+                // repositories _actually contain_ the latest source corresponding to on-chain data (which is subject to
+                // manual syncing itself currently).
```

Comment:
In future we may be able to rely on sources of truth to do this automatically--currently it's not easy, and it's more important to ensure that we don't say anything false, rather than serve outdated sources.